openghg/combine.py

```diff
@@ -0,0 +1,191 @@
+"""
+Combine multiple data objects (subclasses of _BaseData) into one.
+"""
+
+from functools import partial
+import re
+from typing import Any, Callable, cast, Hashable, Mapping, Optional, TypeVar, Union
+
+import numpy as np
+import xarray as xr
+
+from openghg.types import HasMetadataAndData
+
+
+T = TypeVar("T", bound=HasMetadataAndData)  # generic type for classes with .metadata and .data attributes
+
+
+# TODO: review copying vs. mutating data. Is this handled nicely by xarray and dask?
+# TODO: do we need to add options to use dask.delayed like xr.open_mfdataset?
+
+
+def combine_data_objects(data_objects: list[T], preprocess: Optional[Callable] = None) -> T:
+    """Combine multiple data objects with optional preprocessing step.
+
+    Args:
+        data_objects: list of data objects (subclasses of _BaseData, e.g. ObsData or FluxData)
+        preprocess: optional function to call on data objects; must accept type T and return type T
+
+    Returns:
+        single data object with data from each object combined by coordinates, with optional preprocessing applied.
+    """
+    if not data_objects:
+        # without knowing the type T, we can't return a sensible default
+        raise ValueError("`data_objects` must be a non-empty list.")
+
+    if preprocess:
+        data_objects = [preprocess(do) for do in data_objects]
+
+    # combine data by coordinates (this is what is used by xr.open_mfdataset by default; it combines and sorts)
+    datasets = [do.data for do in data_objects]
+    new_data = cast(
+        xr.Dataset, xr.combine_by_coords(datasets)
+    )  # mypy thinks this is an xr.DataArray instead of xr.Dataset
+
+    # combine metadata
+    metadatas = [do.metadata for do in data_objects]
+    new_metadata = metadatas[0].copy()  # TODO: add options for combining metadata
+
+    # make a new data object of same type as input
+    cls = type(data_objects[0])  # need non-empty list of data_objects to get this type
+    result = cls(metadata=new_metadata, data=new_data)
+
+    return result
+
+
+# metadata combination methods
+
+
+def _add_dim_from_metadata(
+    data_object: T,
+    new_dim: str,
+    metadata_key: Optional[str] = None,
+    formatter: Callable = str,
+    drop_metadata_key: bool = True,
+) -> T:
+    """Add a dimension `new_dim `to the data of T using the value corresponding
+    to the key `new_dim` in the metadata of T.
+
+    Args:
+        data_object: object to modify
+        new_dim: name of new dimension to add
+        metadata_key: key of dimension value that will be retrieved from data_object.metadata.
+            If None, then `new_dim` is used as the metadata key.
+        formatter: function to apply to retrieved metadata value. Only applied if value is not None.
+        drop_metadata_key: if True, del `metadata_key` from the metadata of the returned object.
```

Comment:
I added an option to set a replacement value and the default for elevate inlet is now "multiple"

Reply:
Very small thing but at the moment by default am I correct that this now doesn't delete AND doesn't update the metadata (or attribute) value for the relevant key?

Just thinking, we should probably have it that one or other do happen to avoid circumstances with misinformation remaining in the combined data. Perhaps if new_metadata_value is None we could set drop_metadata_key to True?

Reply:
I'll set `drop_metadata_key` to True by default. The new value is added after dropping (if a value is provided) so you don't need to set `drop_metadata_key` to False to use `new_metadata_value`

Reply:
But if we did that then when setting `new_metadata_value` you also have to set `drop_metadata_key` to False otherwise this won't then add the new value.

Reply:
Ah no I see it now - the key will be dropped but then added back if `new_metadata_value` is included - ok great.